File: Inverse Kinematics/src/arm.cpp
#include "arm.hpp"

#include <algorithm>
#include <iostream>
#include <cmath>

static glm::vec2 Lerp(const glm::vec2& start, const glm::vec2& end, float t);

Arm::Arm(int baseX, int baseY, int segCount, int segLength, int segWidth)
	: m_BaseX(baseX), m_BaseY(baseY), m_SegCount(segCount), m_SegLength(segLength), m_SegWidth(segWidth)
{
	// Reduce some ambiguity
	const float defaultAngle = 0.;

	const size_t handIndex = Renderer::LoadTexture("assets/hand.png");
	const size_t armIndex = Renderer::LoadTexture("assets/arm.png");

	m_Segments.resize(m_SegCount);
	m_BaseSeg = new BaseSegment(segLength, 50, defaultAngle, 10, armIndex);
	m_Segments[0] = m_BaseSeg;

	for (int i = 1; i < m_SegCount; i++)
	{
		// Cycle color
		bool isEven = i % 2 == 0;
		SDL_Color color = { 255 - (i * 25), 0 + (i + 25), 255, 255};
		const size_t segIndex = i == m_SegCount - 1 ? handIndex : armIndex;
		const int width = (segIndex == handIndex ? 100 : 50); // Hand needs to be wider than arm

		m_Segments[i] = new Segment(segLength, width,
			defaultAngle, static_cast<uint8_t>(10 - (i / ((m_SegCount / 10) + 1))), segIndex, color);

		m_Segments[i]->AssignParent(m_Segments[i - 1]);
		m_Segments[i - 1]->AssignChild(m_Segments[i]);
	}
}

Arm::~Arm()
{
	for (Segment* seg : m_Segments)
	{
		delete seg;
	}
}

void Arm::Update()
{
	// Point to mouse
	m_Segments.back()->ReverseK(Renderer::mouseX, Renderer::mouseY);
	// Nail to point
	m_BaseSeg->ForwardK();
}

void Arm::Render()
{
	m_BaseSeg->Render();
}

File: Inverse Kinematics/src/python/arm.py
from segment import Segment, BaseSegment
from renderer import Renderer
import pygame

class Arm:
    def __init__(
        self,
        baseX,
        baseY,
        segCount,
        segLengths,
        segWidth,
        floor_offset_px=20.0,
        elbow_up=True
    ):
        self.baseX = baseX
        self.baseY = baseY
        self.segCount = segCount
        self.segWidth = segWidth
        self.floorOffsetPx = float(floor_offset_px)
        self.floorY = float(baseY + self.floorOffsetPx)
        self.elbowUp = bool(elbow_up)

        self.segments = []

        self.baseSeg = BaseSegment(segLengths[0], 50, 0.0, 10)
        self.baseSeg.base_position = (baseX, baseY)
        self.segments.append(self.baseSeg)

        for i in range(1, segCount):
            color = (255 - i * 25, i + 25, 255, 255)
            seg = Segment(
                segLengths[i],
                50,
                90.0,
                max(1, 10 - (i // ((segCount // 10) + 1))),
                color=color
            )
            seg.assign_parent(self.segments[i - 1])
            self.segments[i - 1].assign_child(seg)
            self.segments.append(seg)

    def update(self, target_x=None, target_y=None, speed_scalar=1.0):
        if target_x is None or target_y is None:
            target_x = Renderer.mouseX
            target_y = Renderer.mouseY
        target_y = min(float(target_y), self.floorY)
        self.segments[-1].reverse_k(
            target_x,
            target_y,
            speed_scalar=speed_scalar,
            elbow_up=self.elbowUp
        )
        self.baseSeg.forward_k(floor_y=self.floorY, elbow_up=self.elbowUp)

    def render(self, surface):
        self.baseSeg.render(surface)
        end_effector = self.segments[-1].b
        pygame.draw.circle(
            surface,
            (255, 80, 80),
            (int(end_effector.x), int(end_effector.y)),
            8,
            2
        )

    def get_end_effector_side(self):
        end_effector = self.segments[-1].b
        return (float(end_effector.x), float(end_effector.y))
